Backward compatibility policy (#2879)
ESMValGroup/ESMValTool: doc/sphinx/source/community/backward_compatibility.rst

ESMValTool policy on backward compatibility
===========================================

> V1.0. Agreed with ESMValTool community during October 2022 workshop.

Reviewer: The version number here is slightly confusing to me. Why does this document need to be versioned independently of the rest of the tool?

Author: I agree. Line now removed.

Motivation
----------

Development of recipes or conducting project-related work may require a
rather long period of time during which new versions of the ESMValTool
might become available. For a good user experience and a seamless
workflow, users and developers need to know before upgrading to a new
version if and how their work might be affected (backward
compatibility). This includes, for instance, information about changes
to technical features such as syntax of recipes and configuration files,
and interfaces of shared functions, but also changes that affect the
results of an ESMValTool run, e.g. modification of algorithms or changes
to the order of operators. It is therefore essential that users and
developers have the best advice on how and when to upgrade to new
versions.

While trying to minimise the impact of new code developments on users
and developers by maintaining backward compatibility where possible,
this cannot always be guaranteed. A very restrictive policy might delay
the ESMValTool development and make it more complex for developers to
contribute.

> This document outlines the key principles and proposals of an updated
> ESMValTool policy on backward compatbility.
> 

Reviewer: ```suggestion
This document outlines the key principles and proposals of an updated
ESMValTool policy on backward compatbility.

```
Once merged, this will be the policy, not a proposal

Reviewer 2: Change "compatbility" to "compatibility".

Author: Agreed. "Proposals" now removed and spelling corrected.

-  x indicates a major release
-  y indicates a minor release
-  z indicates a patch release

**Backward-incompatible change:** An ESMValTool change that causes a
recipe to no longer run successfully (a *breaking change*), or which
results in scientifically significant changes in results (a *science
change*).

**Breaking change:** A change which causes a previously working recipe
to no longer run successfully.

**Science change:** A change that alters scientific results. We do not
formally distinguish between trivial science changes (e.g. from changes
in the order of calculations) and more significant changes that would
affect interpretation, although the detail that we communicate will
share any understanding that we have regarding expected impact.

**Benign third-party dependency changes:** A change over which we have
no control, but which we believe will only have trivial technical
impacts (such as a change in font). Such changes are outside of the
scope of this policy, though we will communicate about those we are
aware of.

**Developer of backward-incompatible change:** For the purpose of this
policy, developer is the individual that is responsible for the pull

> request (PR) that is not backward compatible. In the vast majority of
> cases this will be an ESMValCore development.

Reviewer: ```suggestion
request (PR) that is not backward compatible.
```
How do we know that? And does it make the document more clear to add that sentence?

branch of the ESMValTool repository, and can therefore be updated by any
developer in line with the above guidance. Note that the recipe can be
updated by someone other than the original author.

**User recipes:** Recipes developed by any developer outside of the main
branch of the repository (i.e. on a dev/feature branch or outside the
repository completely), and therefore cannot be updated by anyone else.

Scope
-----

The ESMValTool policy on backward compatibility aims at balancing two
competing needs: the occasional need of improvements or maintenance to
break backward compatibility and the need for stability for existing
users and developers. The following aspects are covered by this policy:

-  Key principles and approaches to backward compatibility
-  Guidelines and requirements for *developers of backward-incompatible
   changes*
-  Communication with users and developers about *backward-incompatible
   changes*

Not within the scope of this policy are:

-  Versioning scheme of ESMValTool
-  Breakage of recipes due to changes in input data or dependencies.
   This is covered by the :ref:`broken recipe policy<broken-recipe-policy>`.

> Expectations of developers & users

Reviewer 2: ```suggestion
Expectations of developers, users & funders
```

good timescale for larger changes. For changes that only affect a few
users, shorter timescales could be acceptable. It is also good to note
that we are part of a large software ecosystem (ESMValTool currently
depends on over 500 different software packages), so we may not always
be able to control at what pace changes are made to the software that we
depend upon.

Two-way communication about new and removed features is needed to make
this work. This requires active involvement from both the people
developing the new features and the *recipe developers*. ESMValTool core
developers and ESMValCore core developers need to make sure they clearly
communicate changes. In the first place, this is done by writing good
descriptions in issues and pull requests on GitHub, but some of this
material also makes it to the changelog (where the GitHub pull requests
are linked). It is highly recommended to communicate a relevant
selection (e.g. important new, scheduled for removal, and removed
features) also by other means, to ensure we reach as many people
potentially affected as possible (see “Guidance on handling
*backward-incompatible changes*” section below). We encourage

> involvement of *recipe developers* by organising monthly meetings where

Reviewer 2: Do we really mean to encourage developers to *organize* monthly meetings or rather to *participate* in these meetings?

Reviewer 2: Yes, I like this version much better.

Reviewer 3: "attending" could also be a good word, to get the news you don't have to actively participate. Good spotting though!

everyone is welcome to join, ask questions, and provide feedback.

To meet the needs of users and funders, we should take reproducibility
of older results seriously, but this should not hold us back from
advancing our tools. We can support this by uploading a well tested
container image to an archive that provides a DOI and by providing clear
instructions on how to use such containers.

Helping developers to upgrade
-----------------------------

> *Recipe users* of ESMValTool will be able to successfully run

Reviewer 2: Should we soften "will be able" a bit, e.g. to "should be able"?

Author: Done

*integrated recipes* using a *release*, since all
*backwards-incompatible changes* introduced between *releases* will have
been fixed before the *release* is created. Please note the
:ref:`broken recipe policy<broken-recipe-policy>`.

However, *recipe developers* working on *user recipes* must be provided
with information to enable them to adapt their code to resolve issues
related to *backwards-incompatible changes* when *backwards-incompatible
changes* are introduced to the main branch / when a *release* of
ESMValTool is created.

Guidance on handling *backward-incompatible changes*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As well as helping users to handle *backward-incompatible changes*, the
policy and surrounding tools must help developers avoid making

> *backwards-incompatible changes*. Not many ideas developed on this yet,

Reviewer 3: ```suggestion
*backwards-incompatible changes*. Not many ideas are developed on this yet,
```
incomplete sentence

related to *backwards-incompatible changes* when *backwards-incompatible
changes* are introduced to the main branch / when a *release* of
ESMValTool is created.

Guidance on handling *backward-incompatible changes*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As well as helping users to handle *backward-incompatible changes*, the
policy and surrounding tools must help developers avoid making
*backwards-incompatible changes*. Not many ideas developed on this yet,
but components should include:

-  Testing; *backwards-incompatible changes* should be discovered as
   early in the development process as possible. This motivates
   continued investment in automated testing.
-  Guidance on how to minimise the likelihood of introducing
   *backward-incompatible changes* and how to use deprecation warnings
   when needed (see `developer
   guidance <https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility>`__).
-  Instructions on how to provide the information that *recipe
   developers* need to adapt to the *backward-incompatible change*,
   e.g. the text that will go into *release* notes if we follow the

>    proposal above.

Reviewer: This needs to be clarified/resolved before merging this. Once merged, this isn't a proposal anymore but the actual policy.

Author: It was indeed an unclear point. Is this better: https://github.com/ESMValGroup/ESMValTool/pull/2879/commits/5bc685b158f9cecbe5e9c5b04016e2297b1cd8dd ?

Reviewer: Much better, thanks!

-  General instructions for *recipe developers* working on *user
   recipes* to enable them to adapt their code related to
   *backwards-incompatible changes* (see `ESMValTool_Tutorial: issue
   #263 <https://github.com/ESMValGroup/ESMValTool_Tutorial/issues/263>`__).
-  The developer or reviewer must tag the core development team to
   notify them of the *backward-incompatible change*, and give at least

>    2 weeks for objections to be raised before merging to the main

Reviewer: This also needs to be linked and or mentioned in the following places in the documentation:
- https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#backward-compatibility
- https://esmvaltool--2879.org.readthedocs.build/en/2879/community/review.html#merge

Author: Opened an ESMValCore PR (ESMValGroup/ESMValCore/#1916) for the first one (assuming that an ESMValCore branch requires its own PR)

Author: Added references in both of the above: https://github.com/ESMValGroup/ESMValCore/commit/8af608a679866a2ce479f942e223f872ef66ab08 and https://github.com/ESMValGroup/ESMValTool/commit/14b482df3dddeb0355cc2b78bb617b4b4193f2e5. The first one in particular should be reviewed, as the release notes instructions previously said nothing about a section on backward incompatible changes, and assumed that the release notes will be automated. My change to the instructions introduces manual editing of the ESMValCore release notes.

Reviewer: I added a bit of text along those lines. Approving should be fine because then the author of the pull request knows they are done and can start working on something new, unless anyone objects later.

   branch. If a strong objection is raised the backwards-incompatible
   change should not be merged until the objection is resolved.

> During the *release* process, the following information must be

Reviewer: This section needs to be linked from
- https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#add-release-notes
- https://docs.esmvaltool.org/en/latest/community/release_strategy.html#add-release-notes

Author: Added references in both of the above: https://github.com/ESMValGroup/ESMValCore/commit/8af608a679866a2ce479f942e223f872ef66ab08 and https://github.com/ESMValGroup/ESMValTool/commit/14b482df3dddeb0355cc2b78bb617b4b4193f2e5. The first one in particular should be reviewed (@bouweandela ?), as the release notes instructions previously said nothing about a section on backward incompatible changes, and assumed that the release notes will be automated. My change to the instructions introduces manual editing of the ESMValCore release notes.

provided:

-  **Release notes:** The *release* notes are already documented at

>    https://docs.esmvaltool.org/en/latest/changelog.html, and

Reviewer: ```suggestion
   :ref:`changelog` and :ref:`esmvalcore:changelog` for ESMValTool and ESMValCore respectively, and
```

   “*Backwards-incompatible changes*” is the first section after
   “Highlights”.

   -  **Backwards-incompatible changes:** This section must include
      clear instructions detailing how a *recipe developer* should adapt
      their code for each item in this section, whether the adapted code
      would introduce a *science change*, and the list of affected or
      fixed *integrated recipes* that had to be updated due to the
      *backwards-incompatible changes*, if applicable (to provide
      further examples to *recipe developers* working on *user recipes*
      of how to adapt code).
   -  **Developer guidance:** *Developers* *of backwards-incompatible
      changes* must:

      -  write and include the information required for the
         “*Backwards-incompatible changes*” section in the PR that

>          introduces the *backwards-incompatible change* [action:

Reviewer: This action item needs to be taken care of before merging

Reviewer 2: I added a reference to the backward compatibility policy and some related items to Checklist for pull requests" (`community/code_documentation.rst`).
In addition, I also added a reference to the policy to `doc/contributing.rst` (ESMValCore) in section "Backward compatibility": https://github.com/ESMValGroup/ESMValCore/pull/1918/. I guess that's also covered by @alistairsellar ' PR https://github.com/ESMValGroup/ESMValCore/pull/1916 (just saw this now...). Should we try to merge these to PRs?

@bouweandela, is this roughly what you had in mind?

Author: Thanks @axel-lauer ! I've (manually) merged ESMValGroup/ESMValTool#1916 into ESMValGroup/ESMValTool#1918 and closed the latter, so only one linked PR now.

         guidance related to this needs to be provided in the developer
         documentation]
      -  share details of the *backwards-incompatible change* at the

>          next ESMValTool monthly meeting

Reviewer: ```suggestion
         next monthly ESMValTool community meeting
```
